File: examples/javascript/cloud_browser.js
/**
 * Cloud Browser - Expose sandbox ports and run web services.
 *
 * Usage:
 *   E2B_DOMAIN=localhost:8080 E2B_API_KEY=test-key node cloud_browser.js
 */

const { Sandbox } = require("@e2b/code-interpreter");

async function main() {
  const sandbox = await Sandbox.create({
    apiKey: process.env.E2B_API_KEY || "test-key",
    domain: process.env.E2B_DOMAIN || "localhost:8080",
  });

  try {
    // 1. Write a simple HTTP server
    console.log("1. Writing HTTP server...");
    await sandbox.files.write(
      "/tmp/server.js",
      `
const http = require('http');
const server = http.createServer((req, res) => {
  res.writeHead(200, {'Content-Type': 'application/json'});
  res.end(JSON.stringify({message: 'Hello from sandbox!', status: 'running'}));
});
server.listen(3000, '0.0.0.0');
console.log('Server running on port 3000');
`
    );

    // 2. Start the server
    console.log("2. Starting HTTP server on port 3000...");
    await sandbox.commands.run("node /tmp/server.js &");

    // 3. Get public URL
    console.log("3. Getting public URL...");
    try {
      const host = sandbox.getHost(3000);
      console.log(`   Public URL: ${host}`);
    } catch (e) {
      console.log(`   Port exposure not available: ${e.message}`);
    }

    // 4. Test locally
    console.log("4. Testing locally...");
    const result = await sandbox.commands.run(
      "curl -s http://localhost:3000/ 2>/dev/null || echo 'server not ready'"
    );
    console.log(`   Response: ${result.stdout.trim()}`);
  } finally {
    await sandbox.kill();
    console.log("\nSandbox killed.");
  }
}

main().catch(console.error);

File: examples/python/cloud_browser.py
"""Cloud Browser - Expose sandbox ports and run web services.

Demonstrates port exposure and web service patterns:
1. Create a sandbox
2. Start a web server inside
3. Get the public URL for the port
4. Make requests to the exposed service

Usage:
    export E2B_DOMAIN=localhost:8080
    export E2B_API_KEY=your-api-key
    python cloud_browser.py
"""

import os
import requests
from e2b import Sandbox


def main():
    sandbox = Sandbox.create(
        template="base",
        api_key=os.environ.get("E2B_API_KEY", "test-key"),
        domain=os.environ.get("E2B_DOMAIN", "localhost:8080"),
    )

    try:
        # 1. Install web framework
        print("1. Installing Flask...")
        sandbox.commands.run("pip install flask")

        # 2. Write a web app
        print("2. Writing web app...")
        sandbox.files.write("/tmp/app.py", """
from flask import Flask, jsonify
app = Flask(__name__)

@app.route('/')
def index():
    return jsonify({"message": "Hello from sandbox!", "status": "running"})

@app.route('/health')
def health():
    return jsonify({"status": "healthy"})

if __name__ == '__main__':
    app.run(host='0.0.0.0', port=3000)
""")

        # 3. Start the web server in background
        print("3. Starting web server on port 3000...")
        sandbox.commands.run("python3 /tmp/app.py &")

        # 4. Get the public URL
        print("4. Getting public URL...")
        try:
            host = sandbox.get_host(3000)
            print(f"   Public URL: {host}")

            # 5. Make a request to the exposed service
            print("5. Making request to exposed service...")
            resp = requests.get(f"http://{host}/")
            print(f"   Response: {resp.json()}")
        except Exception as e:
            print(f"   Port exposure not available: {e}")
            # Fallback: test via command
            result = sandbox.commands.run("curl -s http://localhost:3000/ 2>/dev/null || echo 'server not ready'")
            print(f"   Local test: {result.stdout.strip()}")

    finally:
        sandbox.kill()
        print("\nSandbox killed.")


if __name__ == "__main__":
    main()
